File: backend/src/works/entities/work.entity.ts
import { 
  Entity, 
  PrimaryGeneratedColumn, 
  Column, 
  ManyToOne, 
  JoinColumn, 
  CreateDateColumn, 
  UpdateDateColumn, 
  DeleteDateColumn 
} from 'typeorm';
import { UserEntity } from '../../auth/user.entity';

export enum WorkStatus {
  PLANIFICATION = 'En Planificación',
  IN_PROGRESS = 'En Ejecución',
  COMPLETED = 'Completado',
  ARCHIVED = 'Archivado',
}

@Entity('works')
export class WorkEntity {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Column({ type: 'varchar', length: 150 })
  nombre!: string;

  @Column({ type: 'text' })
  direccion!: string;

  @Column({ type: 'date' })
  fechaInicio!: string;

  /**
   * CU-30: fecha final estimada del proyecto, ajustada automáticamente por
   * la ruta crítica (duración total en días desde fechaInicio). Nula hasta
   * el primer recálculo.
   */
  @Column({ name: 'fecha_fin_estimada', type: 'date', nullable: true })
  fechaFinEstimada?: string | null;

  @Column({ type: 'text', nullable: true })
  descripcion?: string; // Opcional (permite valores nulos)

  @Column({ type: 'float', nullable: true })
  latitud?: number;

  @Column({ type: 'float', nullable: true })
  longitud?: number;

  @Column({ 
    type: 'simple-enum', // Clave para compatibilidad polimórfica: PostgreSQL y SQLite (better-sqlite3)
    enum: WorkStatus, 
    default: WorkStatus.PLANIFICATION 
  })
  estado!: WorkStatus;

  // Relación N:1 con Usuario (Propietario) - Garantiza RNF-D-03 (Integridad Relacional)
  @ManyToOne(() => UserEntity, { eager: true, onDelete: 'RESTRICT' })
  @JoinColumn({ name: 'propietarioId' })
  propietario!: UserEntity;

  @Column()
  propietarioId!: string;

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;

  @DeleteDateColumn()
  deletedAt?: Date; // Soporte para RNF-S-03 (Soft-Delete / Audit Log)
}

File: frontend/lib/features/works/domain/entities/work_entity.dart
import 'package:equatable/equatable.dart';

class WorkEntity extends Equatable {
  final String id;
  final String nombre;
  final String direccion;
  final String? descripcion;
  final double? latitud;
  final double? longitud;
  final String fechaInicio;
  final String estado;
  final double progreso;
  final String profesionalId;
  final String? propietarioId;
  // CU-19: datos maestros del propietario vinculado (relación expandida).
  final String? propietarioNombre;
  final String? propietarioEmail;

  const WorkEntity({
    required this.id,
    required this.nombre,
    required this.direccion,
    this.descripcion,
    required this.latitud,
    required this.longitud,
    required this.fechaInicio,
    required this.estado,
    required this.progreso,
    required this.profesionalId,
    this.propietarioId,
    this.propietarioNombre,
    this.propietarioEmail,
  });

  @override
  List<Object?> get props => [
        id,
        nombre,
        direccion,
        descripcion,
        latitud,
        longitud,
        fechaInicio,
        estado,
        progreso,
        profesionalId,
        propietarioId,
        propietarioNombre,
        propietarioEmail,
      ];
}
